core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -116,22 +117,98 @@ class GroupCoordinator(val brokerId: Int,
+    if (notMatch) {
+      responseCallback(joinError(memberId, Errors.INCONSISTENT_GROUP_PROTOCOL))
+    }
+    notMatch
+  }
+
+  private def doUnknownJoinGroup(groupId: String,
+                                 requireKnownMemberId: Boolean,
+                                 clientId: String,
+                                 clientHost: String,
+                                 rebalanceTimeoutMs: Int,
+                                 sessionTimeoutMs: Int,
+                                 protocolType: String,
+                                 protocols: List[(String, Array[Byte])],
+                                 responseCallback: JoinCallback): Unit = {
+    val newMemberId = clientId + "-" + GroupMetadata.generateMemberIdSuffix
+    var group = groupManager.getGroup(groupId).orNull
+    if (group == null) {
       // only try to create the group if the group is not unknown AND
       // the member id is UNKNOWN, if member is specified but group does not
-      // exist we should reject the request
-      groupManager.getGroup(groupId) match {
-        case None =>
-          if (memberId != JoinGroupRequest.UNKNOWN_MEMBER_ID) {
-            responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
-          } else {
-            val group = groupManager.addGroup(new GroupMetadata(groupId, Empty, time))
-            doJoinGroup(group, memberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
-          }
+      // exist we should reject the request.
+      group = new GroupMetadata(groupId, Empty, time)
+      groupManager.addGroup(group)
+    }
+
+    group.inLock {
+      if (group.currentState == Dead) {
+        // if the group is marked as dead, it means some other thread has just removed the group
+        // from the coordinator metadata; this is likely that the group has migrated to some other
+        // coordinator OR the group is in a transient unstable phase. Let the member retry
+        // joining without the specified member id.
+        responseCallback(joinError(newMemberId, Errors.UNKNOWN_MEMBER_ID))
+        return
+      } else if (protocolNotMatch(group, newMemberId, protocolType, protocols, responseCallback)) {
+        return
+      }
 
-        case Some(group) =>
-          doJoinGroup(group, memberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
+      // update the newMemberAdded flag to indicate that the join group can be further delayed
+      if (group.is(PreparingRebalance) && group.generationId == 0)
+        group.newMemberAdded = true
+
+      // If member id required, register the member with unknown metadata,
+      // and send back a response to call for another join group request with allocated member id.
+      if (requireKnownMemberId) {
+        group.addUnknownMember(newMemberId, group.groupId, protocolType, protocols)
```

Comment:
I am wondering if we should call `completeAndScheduleNextHeartbeatExpiration`. This would ensure that the pending member will eventually get timed out if the consumer never rejoins.

Reply:
I also think we might want to call it - that call should satisfy the KIP's statement:

>  We also handle the registered member id eviction through session timeout so that the pre-allocation map will not grow indefinitely, although the map size should be trivial since we only store a random generated id.

right?

Reply:
When I start the implementation, I actually realize that it's probably simpler just to keep the `pendingMembers` map as a simple check without enforcing purgatory. The reason is that the size of the map is trivial and will be evicted after rebalance timeout, which I feel could be more clean and efficient (saving purgatory space in the meantime). What do you think?

Reply:
The rebalance timeout is client-controlled, right? It's in the JoinGroup request and it comes from `max.poll.interval.ms` as far as I know. Users can set that to Integer.MAX_INT (as they have with that original JIRA that spawned the discussions and KIP-389) and fill this up to infinity (at least it's just a string and not MemberMetadata)

In my understanding, that's what we're partly fixing with this PR.

Reply:
Yep, +1. It's necessary to handle the rebalance timeout case here, thanks for pointing out!